### PR TITLE
fix: validate untrusted input before indexing, asserting or dereferencing

### PR DESCRIPTION
**Pull request by fbsobreira**
Wave 1 of the repository review: twelve findings that share one defect class — data from the node,
from a keystore file, or from a caller reaching an index, a type assertion, or a pointer dereference
with no check.

## `pkg/client` — node responses

| | |
|---|---|
| **C8** | `ParseTRC20StringProperty` took the ABI string length from contract return data and evaluated `2*int(l)`, which overflows to a negative value, passed the bounds check, then panicked on the slice. |
| **W11** | Four TRC20 queries indexed `GetConstantResult()[0]` unguarded; a node may report success with no entries. |
| **W12** | `TRC20Send`/`Approve`/`TransferFrom` encoded amounts with `LeftPadBytes`. |
| **W27** | `DeployContractCtx` assigned through `tx.Transaction.RawData` without checking the result code. |
| **W28** | `UpdateAccountPermissionCtx` made eight unchecked assertions on caller-supplied maps. |
| **W43** | `GetTransactionInfoByIDCtx` dereferenced a possibly-nil response. |
| **W44** | `WithdrawExpireUnfreeze`, `DelegateResource`, `UnDelegateResource` skipped the result-code check every sibling performs. |

### C8 had two failure modes, not one

Verified by running the original bounds check against each input:

| Declared length | Old behaviour |
|---|---|
| 2^62 | **panic** — `slice bounds [:9223372036854775936]` |
| 2^63 | silent `""` — `2*l` wraps to exactly 0 |
| max uint64 | **panic** |
| 2^254 | silent `""` — `Uint64` truncates to the low bits |
| max uint256 | **panic** |

A hostile `name()` could therefore return an empty token name **with a nil error**, not only crash. Now
`IsUint64` catches the truncation and the bound is expressed as a division, so no multiplication exists
to overflow.

### W12 encoding is unchanged for valid amounts

All three sites now use `trc20enc.PadUint256`. Confirmed **byte-identical** to `LeftPadBytes` across
`0`, `1`, `1e6`, max uint64, 2^255−1 and max uint256 — zero mismatches, so calldata for any amount that
worked before is untouched. What changes is that `nil` (panicked), negative (silently became a real
positive transfer, because `big.Int.Bytes` returns the absolute value) and >256-bit values (appended
whole, because `LeftPadBytes` does not truncate) are now rejected.

## Five further sites of the same class

Found while reviewing the above. `triggerConstantContract`, `triggerContract`, `DeployContract` and
their callers in `contracts.go`, plus `TRC20CallCtx`, all read `tx.Result.Code` **directly** rather
than through the nil-safe accessor, so a response with no `Result` panicked. `TRC20CallCtx` gates all
four W11 queries — the constant-result indexing was guarded while a nil-`Result` panic sat upstream on
the same path. `triggerContract` also assigned through `tx.Transaction.RawData` with no nil check, the
same shape as W27.

## `pkg/common/numeric`

- **W37** — unanchored regex, so `"1e5x"` matched, `Atoi("5x")` failed silently, and the function
  returned `1` with a **nil error**. Anchored; both errors propagate.
- **W38** — a nil `*big.Int` from `SetString` reached `Mul` and panicked. It also accepted a sign
  (because `big.Int.SetString` does), yielding a *negative* `Dec` out of a hex parser.

## `pkg/keystore`

- **W10** — `getKDFKey` asserted `salt`/`prf` to string and every numeric parameter to `float64` with no
  comma-ok, and passed `n`, `r`, `p` to scrypt with **no upper bound** — all *before* the MAC is
  verified. A crafted file could panic the process or force an unbounded derivation. Bounds leave room
  for `StandardScryptN` (256 MiB working set) and pbkdf2 defaults.
- **W26** — `NewKeyForDirectICAP` retried until an address began with `0x00`, which a TRON address never
  does, so every call recursed until the stack overflowed. Removed; ICAP has no meaning on TRON.

## Testing

**W66** adds the negative-case coverage whose absence let W37 and W38 survive.

Tests assert both that input is rejected **and** that nothing panics. Branches a real gRPC round trip
cannot produce — a nil `TransactionInfo`, a nil `Result` — are reached by stubbing the exported `Client`
field, which is also how an SDK consumer can produce them. (`grpc-go` materialises a `(nil, nil)` server
return into an empty message, so a mock server cannot reach those guards; coverage confirmed the branch
was dead before this change.)

Every touched function keeps its pre-existing happy-path test. Coverage: `pkg/client` 83.5 → 84.7%,
`numeric` 91.7 → 92.9%, `keystore` 84.0 → 86.1%.

`make lint` (0 issues) · `make test` (20/20 packages) · `go test ./cmd/...` · `make` — all clean.

## Not included

**W42** (nil `g.Client` before `Start`) is a mechanical sweep of **80 call sites across 13 files** and
does not belong in the same diff as these targeted guards. It gets its own PR.

## ⚠️ Behaviour change for release notes

**W27 and W44 change observable behaviour.** Callers that previously received a `TransactionExtention`
for a request the node *rejected* now receive an error. That is the fix — the failure previously
surfaced only after signing and broadcast, or not at all — but it is breaking for anyone who was
ignoring the result code.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved detection and reporting of node-rejected transactions across account, asset, bank, contract, exchange, proposal, resource, transfer, and witness operations.
  * Prevented crashes on empty, nil, or malformed responses, permissions, transaction lookups, and TRC20 calls.
  * Hardened numeric parsing and TRC20 decoding against overflow, underflow, and malformed input.
  * Strengthened keystore decryption by validating cipher, KDF, and encryption parameters.
* **Breaking Changes**
  * Removed the Direct ICAP key-generation helper.
* **Tests**
  * Added extensive regression coverage for malformed inputs and rejected responses.

### CODE REVIEW

**Comment by fbsobreira** [comment]
All three addressed in `3fe12842`. Each was verified against the code before fixing — one did not reproduce as described, so noting that explicitly.

### ✅ Exponent bound — confirmed, and worse than reported

Real, and it is a hard crash from a plain string. `NewDecFromString("1e-9223372036854775808")`:

```
runtime: goroutine stack exceeds 1000000000-byte limit
fatal error: stack overflow
```

Your analysis of the mechanism is exactly right: `Pow` negates a negative exponent and `math.MinInt` stays negative under negation.

There is a second overflow you did not mention, which I hit because my first bound was still too loose. I picked ±1000 and my own test case `1e1000` panicked with `Int overflow` — `Dec` is capped at `255+DecimalPrecisionBits` bits over an 18-decimal scale. So the safe range is much narrower. Measured it rather than computed it:

```
largest safe positive exponent: 76 (10^76 ok, 10^77 panics)
```

Bounded to ±76. The original input now returns `exponent -9223372036854775808 in "1e-9223372036854775808" outside [-76, 76]`.

### ⚠️ `dklen` — the panic does not reproduce, but fixed anyway

I could not reproduce the slice-bounds panic. A keystore with `"dklen": 1` returns a clean `could not decrypt key with given passphrase`, not a crash.

`derivedKey[16:32]` is a **slice expression**, so it is legal whenever `cap ≥ 32` — not `len ≥ 32`. Both KDFs round up to a whole hash block internally, so they return `cap 32` even for `dklen 1`:

```
dklen= 1  scrypt len=1  cap=32 panics=false | pbkdf2 len=1  cap=32 panics=false
dklen=15  scrypt len=15 cap=32 panics=false | pbkdf2 len=15 cap=32 panics=false
dklen=31  scrypt len=31 cap=32 panics=false | pbkdf2 len=31 cap=32 panics=false
```

So it is not Critical and not a panic. But the underlying point stands: reading past `len` into spare capacity is fragile, it depends on an implementation detail of `golang.org/x/crypto`, and a sub-32 `dklen` is invalid per the V3 spec regardless. Minimum is now 32, which removes the dependency. `EncryptDataV3` always writes `scryptDKLen = 32`, so round trips are unaffected.

### ✅ Non-nil result with error — confirmed

Correct on both counts. `TRC20SendCtx`, `TRC20ApproveCtx` and `TRC20TransferFromCtx` all `return g.TRC20CallCtx(...)` directly, so the rejection payload reaches the caller. Returns `nil` now, matching `triggerContract` and `DeployContractCtx`.

Worth flagging for release notes: a caller who inspected the returned extention on error to read a revert reason from a constant call loses that. No in-tree caller does — all four query methods return early on `err != nil`.

### Tests

- exponent: `math.MinInt`, `math.MaxInt`, `±1e5` overflow cases, `1e77`/`1e-77`, plus `1e76` as a valid boundary case
- dklen: `1`, `31`, and a pbkdf2 `16`

`make lint` 0 issues · `make test` 20/20 · `go test ./cmd/...` · `make` — all clean.

**Comment by fbsobreira** [comment]
Both confirmed and fixed in `9477ab60`. The magnitude point was right and I had it wrong twice.

### `9e76` — confirmed

The exponent bound I added was necessary but **not sufficient**, exactly as you said. Magnitude depends on the mantissa too:

```
1e76   -> ok        5e76   -> ok        1.5e76 -> ok
9e76   -> PANIC: Int overflow
99e75  -> PANIC: Int overflow
```

### Hex — confirmed, and worse than the report

You flagged "lengths that would overflow `Pow`". Measured, the threshold is **64 characters** — an ordinary ABI `uint256` word:

```
16 chars -> ok
32 chars -> ok
64 chars -> PANIC: Int overflow
200 chars -> PANIC: Int overflow
```

So the most natural input to a hex→decimal helper crashed.

### Fix

I tried computing the limit in the parser twice and got it wrong both times — first `±1000` (my own test case `1e1000` panicked), then `±76`, which `9e76` walked straight through. Reproducing `Dec`s `255+DecimalPrecisionBits` cap outside `Dec` is the wrong approach.

Both parsers now recover the documented `"Int overflow"` panic and convert it to `ErrOutOfRange`, re-raising anything else so real bugs still surface. That covers every arithmetic path exactly instead of approximating it.

The exponent bound stays, for the one case recover cannot help with: `math.MinInt` reaching `Pow`, where negating the minimum signed int leaves it negative and it recurses until the stack overflows.

```
9e76        -> error: value out of representable range
99e75       -> error: value out of representable range
64 chars    -> 0.000000000000000000
1e76, 5e76, 1.5e76, 32-char hex -> still parse correctly
```

`NewDecFromHex` keeps its signature and degrades to zero, consistent with how it already reports other invalid input.

One correction to my own test: I initially added `12345e72` as an overflow case. It is `1.2345e76`, which is representable — the test caught it and it is removed.

`make lint` 0 issues · `make test` 20/20 · `go test ./cmd/...` · `make` — all clean.

File: pkg/client/txext.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// or a non-zero result code with no Transaction. Builders also reject SUCCESS with
+// a missing body so callers never receive a hollow extension to sign (custom
+// WalletClient stubs and partial responses can produce that shape).
+//
+// op is a short name used in the missing-transaction error (e.g. "delegate resource").
+func requireTxExtension(tx *api.TransactionExtention, op string) error {
+	if tx == nil || proto.Size(tx) == 0 {
+		return fmt.Errorf("bad transaction")
+	}
+	if code := tx.GetResult().GetCode(); code != 0 {
```
[the remark this answers is not in the document]
Comment (fbsobreira) [reply pkg/client/txext.go:31]: Fixed — `requireTxExtension` now rejects a nil `Result` before calling `GetCode()` (which returns `SUCCESS` for a nil `*Return` as a protoc-gen-go zero-value default, not a node-reported success). Full nodes we checked always set `Result`; this blocks hollow responses that only carry `RawData`. Added a regression test for RawData-with-no-Result.

File: pkg/keystore/passphrase.go [hunk 2]
```diff
@@ -270,6 +260,53 @@ func DecryptDataV3(cj CryptoJSON, auth string) ([]byte, error) {
...
+func decodeCipherFields(macHex, ivHex, cipherHex string, cbc bool) (mac, iv, cipherText []byte, err error) {
+	if len(macHex) > macLen*2 {
+		return nil, nil, nil, fmt.Errorf("crypto: mac hex length %d exceeds limit %d", len(macHex), macLen*2)
+	}
+	if len(ivHex) > aesIVLen*2 {
+		return nil, nil, nil, fmt.Errorf("crypto: iv hex length %d exceeds limit %d", len(ivHex), aesIVLen*2)
+	}
+	if len(cipherHex) > maxCiphertextLen*2 {
+		return nil, nil, nil, fmt.Errorf("crypto: ciphertext hex length %d exceeds limit %d", len(cipherHex), maxCiphertextLen*2)
+	}
```
[the remark this answers is not in the document]
Comment (fbsobreira) [reply pkg/keystore/passphrase.go:282]: Fixed — `EncryptDataV3` now enforces the same `maxCiphertextLen` (1024) as `DecryptDataV3` before KDF work, so oversized plaintexts cannot encrypt into a `CryptoJSON` this package refuses to decrypt. Documented on the function and added boundary tests (at limit / one over).
